crates/configuration/src/configuration.rs

```diff
@@ -1,85 +1,339 @@
 use std::{collections::BTreeMap, path::Path};
 
-use anyhow::ensure;
+use anyhow::{anyhow, ensure};
 use itertools::Itertools;
-use schemars::JsonSchema;
-use serde::Deserialize;
+use mongodb_support::BsonScalarType;
+use ndc_models as ndc;
 
-use crate::{native_procedure::NativeProcedure, read_directory, schema::ObjectType, Schema};
+use crate::{
+    native_procedure::NativeProcedure,
+    native_query::{NativeQuery, NativeQueryRepresentation},
+    read_directory, schema, serialized,
+};
 
-#[derive(Clone, Debug, Default, Deserialize, JsonSchema)]
-#[serde(rename_all = "camelCase")]
+#[derive(Clone, Debug, Default)]
 pub struct Configuration {
-    /// Descriptions of collections and types used in the database
-    pub schema: Schema,
+    /// Tracked collections from the configured MongoDB database. This includes real collections as
+    /// well as virtual collections defined by native queries using
+    /// [NativeQueryRepresentation::Collection] representation.
+    pub collections: BTreeMap<String, ndc::CollectionInfo>,
+
+    /// Functions are based on native queries using [NativeQueryRepresentation::Function]
+    /// representation.
+    pub functions: BTreeMap<String, ndc::FunctionInfo>,
+
+    /// In query requests functions and collections are treated as the same, but in schema
+    /// responses they are separate concepts. So we want a set of [CollectionInfo] values for
+    /// functions for query processing, and we want it separate from `collections` for the schema
+    /// response.
+    pub function_collection_infos: BTreeMap<String, ndc::CollectionInfo>,
```

Comment:
Will `functions` and `function_collection_infos` have the same set of keys?
If so, would it be better to represent this as a `BTreeMap<String, (ndc::FunctionInfo, ndc::CollectionInfo)>`?

Reply:
That does make sense. But I think it's a matter of style. There isn't any case where we want to look up both `FunctionInfo` and `CollectionInfo` - the schema handler wants one, the query handler wants the other. We shouldn't be constructing `Configuration` values without going through `validate` except in tests so I'm not concerned about accidentally created an invalid state.

Reply:
Ok I made this change. It does make the validate function simpler. But I'd rather not combine the procedure maps since I like keeping the ndc stuff somewhat separated from the mongodb-specific stuff.